libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "critical_section.h"
+#include "interrupt.h"
+#include "status.h"
+#include "stm32f0xx_interrupt.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct Timer {
+  uint32_t duration;
+  void *context;
+  bool inuse;
+  SoftTimerCallback callback;
+} Timer;
+
+static volatile Timer s_soft_timer_array[SOFT_TIMER_MAX_TIMERS];
+static volatile SoftTimerID s_active_timer_id;
+static volatile uint32_t s_last_update_count;
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
```

Comment:
So I won't be able to fully reply about the timers until after work tomorrow, but the basic goal is to minimize processing. I think there are two parts to this:
A) using a free-running counter and precomputing timers based on the current counter value 
B) separating the free-running counter from an interrupt-driven timer 

I will be honest, I didn't realize there was only one 32-bit timer so accomplishing part B may require some more thinking. :P The idea was that stopping and starting the timer would cause it to be offset by a number of microseconds, especially if you think about the speed of the system clock. If we're running at 48Mhz (I'm pretty sure we're currently at 8Mhz), we only get 48 cycles per microsecond. That's not a lot of time to loop through all the timers without losing accuracy. With a free-running counter, your interrupt might not always fire exactly on time, but you'll know how off you are. Imagine if your timer interrupt was actually preempted by another interrupt for 1ms. If you disabled your timer, you'd never know how close to your target time you were actually able to get. 

I think part A can be done with your current implementation though. That was the idea behind using a free-running counter, because right now it's pretty much the same thing as what you had before. I see some computation, but I also see a loop that iterate over every single timer and updates its duration. Ideally, you'd be able to just compute the expected value once and just check when the interrupt fires. If you don't need to recompute the expected duration each time, you should also be able to optimize the ordering of your timers so you don't need to loop through all of them. We can probably guess that most new timers will be executed in the back portion of the current active timers.

I am a little concerned about how we handle callbacks from a soft timer, since they could potentially take a long time, but I don't know the best way to handle that.

Just to clear, it's awesome that you have this working, I'm just wondering if it's possible to make it better. :) Realistically, we'll probably only need it to be accurate within 300us or something like that.